File: 백준/Gold/16398. 행성 연결/행성 연결.cc
//https://www.acmicpc.net/problem/16398
#include <bits/stdc++.h>
using namespace std;

int n;
using ll = long long;

vector<int> p(1001, -1);



int find(int v)
{
	if (p[v] <0)
		return v;

	return p[v] = find(p[v]);
}

bool is_diff_group(int u, int v) {
	u = find(u); v = find(v);
	if (u == v) return 0;
	if (p[u] == p[v]) p[u]--;
	if (p[u] < p[v]) p[v] = u;
	else p[u] = v;
	return 1;
}

vector<tuple<int, int, int>> edges;

int main()
{
	ios::sync_with_stdio(0);
	cin.tie(0);

	cin >> n;

	for (int i = 1; i <= n; i++)
		for (int j = 1; j <= n; j++)
		{
			int cost;
			cin >> cost;
			if (i > j)
				edges.push_back({cost, i, j });
		}

	sort(edges.begin(), edges.end());

	int cnt = 0;
	ll ans = 0;

	
	for (auto edge : edges)
	{
		int a, b, cost;
		tie(cost, a, b) = edge;

		if (!is_diff_group(a, b))
			continue;
		ans += cost;
		cnt++;
		if (cnt == n - 1)
			break;
	}

	cout << ans;


}

File: 백준/Gold/16398. 행성 연결/행성 연결.java
import java.util.*;
import java.io.*;

public class Main{
	public static void main(String[] args) throws IOException{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		N = Integer.parseInt(br.readLine());
		adj = new long[N+1][N+1];
		
		uf = new int[N+1];
		for(int y= 1; y<=N; y++) {
			uf[y] = y;
			st = new StringTokenizer(br.readLine());
			for(int x= 1; x<=N; x++) {
				adj[y][x] = Long.parseLong(st.nextToken());
				
				if(adj[y][x] != 0) {
					edges.add(new Edge(y,x,adj[y][x]));
				}
			}
		}
		
		Collections.sort(edges);
		int cnt = 0;
		long ans = 0;
		for(Edge edge : edges) {
			if(union(edge.u, edge.v)) {
				ans += edge.cost;
				cnt ++;
				
				if(cnt == N-1)
					break;
			}
		}
		
		System.out.println(ans);
	}
	
	static int find(int x) {
		if(uf[x] == x)
			return x;
		
		return uf[x] = find(uf[x]);
	}
	
	static boolean union(int x, int y) {
		x = find(x);
		y = find(y);
		
		if(x == y)
			return false;
		
		uf[y] = x;
		return true;
	}
	
	static int uf[];
	static List<Edge> edges = new ArrayList<>();
	static StringTokenizer st;
	static int N;
	static long adj[][];
	static class Edge implements Comparable<Edge>{
		int u,v;
		long cost;
		public Edge(int u, int v, long cost) {
			this.u = u;
			this.v = v;
			this.cost = cost;
		}
		
		public int compareTo(Edge o) {
			return Long.compare(this.cost, o.cost);
		}
		
		public String toString() {
			return String.format("%d %d %d", u,v,cost);
		}
	}
}

